dhis-2/dhis-support/dhis-support-db-migration/src/main/resources/org/hisp/dhis/db/migration/2.39/V2_39_14__Add_referral_column_to_relationshiptype.sql

```diff
@@ -1,6 +1,2 @@
-ALTER TABLE relationshiptype ADD COLUMN IF NOT EXISTS referral BOOLEAN;
-
-UPDATE relationshiptype SET referral = FALSE WHERE referral IS NULL;
-
-ALTER TABLE relationshiptype ALTER COLUMN referral SET NOT NULL;
+ALTER TABLE relationshiptype ADD COLUMN IF NOT EXISTS referral BOOLEAN DEFAULT false;
```

Comment:
A reason for doing it the way it was before would be if there is a chance the column exists. It would then still make sure to update the null values to false and add the not null constraint.

So now it would not do the latter steps. If you also want to account for that fact, you would need to adapt it.

@ameenhere maybe you have some more thoughts on this as you've seen a few migrations that failed in the last months I think 😬

Reply:
Shouldn't we still set the NOT NULL constraint on the column even though we set the default?

The column can otherwise still contain null values.

Reply:
Although practically it may not be possible to update it to null through hibernate as booleans are primitives. But it does leave a small window open to either have null values via jdbctemplate or directly updating in the db.

Apart from that tiny window, the script is idempotent and looks good to me.

Reply:
Look at the suggested change that I put.
I think that to be really safe we need to make sure we remove null values (as @teleivo said, the column could already be present and allow null values, if we just add if not exist, it will be skipped and null values would still be allowed) and we need to set the not null to be sure it cannot be manually changed

Reply:
It will not fail but it will not be applied if the column exists

Reply:
Ah good point.